sessionization/sessionization.md

```diff
@@ -0,0 +1,597 @@
+        axisFormat  %y
+        todayMarker off
+
+        section Key 1
+        [33] :1,12
+        [2] :14,15
+      ```
+
+
+The code below implements the algorithm described above. As mentioned above, you can find the full, feature-complete implementation on [<img src="github.png" style="height:16px;"/> apache/beam](https://github.com/apache/beam/blob/3a4d57eb8976c5f503b32d478a80b1800490f66f/runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/Aggregators.java#L188-L228).
+
+```java
+private final Combine.CombineFn<InputT, AccumT, OutputT> fn;
+
+@Override
+public TreeMap<IntervalWindow, AccumT> reduce(TreeMap<IntervalWindow, AccumT> buff, WindowedValue<KV<KeyT, InputT>> input) {
+  // for each input window
+  for (IntervalWindow window : (Collection<IntervalWindow>) input.getWindows()) {
+    AccumT acc = null;
+    IntervalWindow first = null, last = null;
+    // start with window before or equal to new window (if exists)
+    Map.Entry<IntervalWindow, AccumT> lower = buff.floorEntry(window);
+    if (lower != null && window.intersects(lower.getKey())) {
+      // if intersecting, init accumulator and extend window to span both
+      acc = lower.getValue();
+      window = window.span(lower.getKey());
+      first = last = lower.getKey();
+    }
+    // merge following windows in order if they intersect, then stop
+    for (Map.Entry<IntervalWindow, AccumT> entry : buff.tailMap(window, false).entrySet()) {
+      AccumT entryAcc = entry.getValue();
+      IntervalWindow entryWindow = entry.getKey();
+      if (window.intersects(entryWindow)) {
+        // extend window and merge accumulators
+        window = window.span(entryWindow);
+        acc = acc == null ? entryAcc : fn.mergeAccumulators(ImmutableList.of(acc, entryAcc));
+        if (first == null) {
+          // there was no previous (lower) window intersecting the input window
+          first = last = entryWindow;
+        } else {
+          last = entryWindow;
+        }
+      } else {
+        break; // stop, later windows won't intersect either
+      }
+    }
+    if (first != null && last != null) {
+      // remove entire subset from from first to last after it got merged into acc
+      buff.navigableKeySet().subSet(first, true, last, true).clear();
+    }
+
+    if (acc == null) {
+      // init accumulator if necessary
+      acc = fn.createAccumulator();
+    }
+    // add input and set accumulator for new (potentially merged) window
+    buff.put(window, fn.addInput(acc, input.getValue().getValue()));
+  }
+  return buff;
+}
+```
+
+
+##### Merging aggregation buffers
+
+The idea to merge our tree buffers is simple; we traverse both trees in order and always pick the next smallest window (in time). While it intersects a previous one, we expan the windows and merge the accumulators, if not we add it to a new tree buffer and continue until there's no more windows; once the final window is added we're done.
```

Comment:
I'd say "in timestamp" instead of "in time" to avoid confusion with the window width